### DISCUSSION TITLE
Debug test with VSCode

### DISCUSSION THREAD

**Discussion by ethancadoo** [fzyzcjy/flutter_convenient_test#264] [Q&A]
I have a failing integration test that I would like to debug within vscode.
If I try to create a launch configuration like so:
**launch.json**
```
   {
      "name": "convenient test",
      "request": "launch",
      "type": "dart",
      "program": "integration_test/app_test.dart",
      "args": [
        "--dart-define=ENV=dev",
        "--flavor",
        "development",
        // "--host-vmservice-port=9753", //This was causing the error 'unknown' arg when launching with vscode
        "--disable-service-auth-codes",
        "--dart-define=CONVENIENT_TEST_APP_CODE_DIR=${workspaceFolder}",
        "--dart-define=CONVENIENT_TEST_MANAGER_HOST=10.0.2.2"
      ],
    },
```
I expect vscode to automatically attach a debugger but instead I get the error:

`No tests were found.`

When I try to start the client from the command line and attach the vscode debugger afterwards I get the following errors in the client logs:
`E/flutter (28117): [ERROR:flutter/runtime/dart_isolate.cc(143)] Could not prepare isolate.`
`E/flutter (28117): [ERROR:flutter/runtime/runtime_controller.cc(382)] Could not create root isolate.`

And the manager is never able to start the test. The manager is able to cause a hot reload.

**launch.json** for attaching to running process.
```
   {
      "name": "Dart: Attach to Process",
      "type": "dart",
      "request": "attach"
    },
```

Is this a limitation of this package or am I doing something wrong?

**Comment by fzyzcjy** [comment]
Do you successfully debug naive flutter integration tests? If not maybe need to do that first.

**Reply by ethancadoo** [reply]
Is there a way to invoke the convenient tests with a naive integration test driver?
All my tests are written with the tools supplied by this package.

**Reply by fzyzcjy** [reply]
Well, my question is, I am not sure whether a naive flutter integration test (without this package) *supports* debugging. If not, this package may not as well (though there are quite a few differences).

Btw personally speaking, I love to do logging to see bugs, instead of attaching a debugger

**Reply by ethancadoo** [reply]
I can setup a test if need be but it seems like it is possible https://stackoverflow.com/questions/52024702/flutter-debugging-an-integration-test

Vscode also auto scans projects for tests and tries to give you specific entry points into them with the "Unit Test" tab. 
It would be great if tests written for this package could be be compatible with this existing tooling.

**Reply by fzyzcjy** [reply]
I see. This package does not have weird things - it only runs the worker with `flutter run` so I guess it should work. Feel free to explore why it does not!

**Reply by marcel-ploch** [reply]
Hey together,

today I found out that you need to put the test files in a separate folder:

```
{
            "name": "convenient_test_android",
            "request": "launch",
            "type": "dart",
            "args": [
                "--flavor",
                "dev",
                "--disable-service-auth-codes",
                "--dart-define",
                "CONVENIENT_TEST_MANAGER_HOST=10.0.2.2",
                "--dart-define",
                "CONVENIENT_TEST_APP_CODE_DIR=lib/main_dev.dart -d 'emulator-5554'",
                "-v"
            ],
            "vmAdditionalArgs": [
                "--host-vmservice-port=9753",
            ],
            "program": "conv_test/main_test.dart"
        },
```

Also nice is this Extension to the launch json
```
"codeLens": {
                "for": [
                    "run-file",
                    "debug-file"
                ],
                "path": "conv_test",
                "title": "${debugType} Conv Test"
            },
```
With this CodeLens will add fitting Labels to start the test.

If the test files are under `conv_test` the VS Code extension is not starting the process with flutter test it is starting with flutter run and than it is doing with charm :)

Kind regards

Marcel Ploch
